### DISCUSSION TITLE
Markdown / Markdown Preview improvements  - Tracking Issue

### DISCUSSION THREAD

**Discussion by nathansobo** [zed-industries/zed#23951] [Feature Requests]
[All Markdown issues](https://github.com/zed-industries/community/issues?q=is%3Aopen+is%3Aissue+label%3Amarkdown+sort%3Areactions-%2B1-desc)

## LSP

Investigate this, this may solve multiple other requests

- [ ] https://github.com/zed-industries/zed/issues/5243

## Syntax highlighting

- [x] https://github.com/zed-industries/zed/issues/5217
- [ ] [`_` breaks highlighting](https://github.com/zed-industries/zed/issues/5346)
- [ ] https://github.com/zed-industries/extensions/issues/203
- [x] https://github.com/zed-industries/zed/issues/5761

## Render / Embed

- [ ] Render bold text bold, italic text italic, etc.
- [ ] Document embedding via multibuffers ([@Kethku's suggestion](https://github.com/zed-industries/zed/issues/1679#issuecomment-1265733773))
- [x] https://github.com/zed-industries/zed/issues/992
- [ ] https://github.com/zed-industries/zed/issues/5087
- [x] https://github.com/zed-industries/zed/issues/5313
- [ ] https://github.com/zed-industries/zed/issues/10022
- [ ] https://github.com/zed-industries/zed/issues/10696
- [x] https://github.com/zed-industries/zed/issues/11024
- [x] https://github.com/zed-industries/zed/issues/13246
- [ ] https://github.com/zed-industries/zed/issues/13603
- [ ] https://github.com/zed-industries/zed/issues/15379
- [x] https://github.com/zed-industries/zed/issues/18540
- [x] https://github.com/zed-industries/zed/issues/18542
- [x] https://github.com/zed-industries/zed/issues/19061
- [ ] https://github.com/zed-industries/zed/issues/19157
- [ ] https://github.com/zed-industries/zed/issues/21319
- [ ] https://github.com/zed-industries/zed/issues/23218

## Autocomplete / Autoformatting

- [ ] https://github.com/zed-industries/zed/issues/5089
- [x] https://github.com/zed-industries/zed/issues/5382
- [x] https://github.com/zed-industries/zed/issues/5226

## Misc

- [x] https://github.com/zed-industries/zed/pull/21630
- [ ] https://github.com/zed-industries/zed/issues/4791
- [ ] https://github.com/zed-industries/zed/issues/4924
- [ ] https://github.com/zed-industries/zed/issues/5305
- [ ] https://github.com/zed-industries/zed/issues/14115
- [ ] Support for Emoji Shortcodes

## HTML

- [ ] https://github.com/zed-industries/zed/issues/14830
- [ ] https://github.com/zed-industries/zed/issues/21992
- [ ] https://github.com/zed-industries/zed/issues/22056

**Comment by iamnbutler** [comment]
I'd love for us to detect checklists as well (`- [ ]`) and add a new one when making a new line.

**Reply by human85** [reply]
agree

**Comment by fdionisi** [comment]
Not sure if the following should be added here on in a separate bug ticket.

When adding italic styling using `_` to the last word of a sentence on the last line of the editor, syntax highlight is not applied until you enter a new line below.

https://user-images.githubusercontent.com/8927326/181229411-51626d9e-ed0f-474c-9bfc-df0a89cbfb32.mp4

**Comment by srl295** [comment]
- should support `backticks` (`` `backticks` ``) also

(didn't seem to be exactly covered by the above.)

**Comment by bbkane** [comment]
Inline image rendering would make a HUGE impact on how I write Markdown (it's currently the main reason I use https://typora.io/ ) (see [my blog](https://www.bbkane.com/blog/how-i-take-notes/) for more ~~ranting~~ information about what I need to take notes)

**Comment by red-co** [comment]
Can we add support for id anchor jumps in Markdown, // If possible, allow this specific syntax to jump directly into the editor in any source code comment block.    

Some editors allow setting ids or aliases for headers.    
    
It may be implemented with html like `<a id="anchor">name</a>`, but this hack only works for page html, and editor inside  jumps do not support it.

**Comment by dimitre** [comment]
I've noticed now Zed is trimming down double spaces in the end of line (Markdown code for line break) on file save. 
it would be great to be able to keep them in .md files

**Reply by mtalexan** [reply]
The trimming of trailing whitespace is a formatter setting in Zed.  I personally have it disabled globally just for this reason, but maybe there's a way to disable it only for the Markdown language files?  I'm not particularly clear which options are available to configure per-language.

**Reply by calebmeyer** [reply]
Personally, I don't like using invisibles (even if they are rendered on selection or by default) to indicate anything semantically important. I use `<br />` tags if I really need a markdown line break, and that lets me keep the "trim trailing whitespace on save" option on.

That said, I am in support of having the option for people who do prefer double space line breaks.

**Comment by squingo44** [comment]
Default Markdown highlighting doesn't colour headers, which isn't the behaviour I expect coming from other editors:

Zed:
<img width="1425" alt="Screenshot 2025-05-26 at 14 45 16" src="https://github.com/user-attachments/assets/75baa18c-ab98-45b4-b28c-4ffd79f3f97d" />

VSCode:
<img width="871" alt="Screenshot 2025-05-26 at 14 37 41" src="https://github.com/user-attachments/assets/76c29dc5-9119-4a39-9585-c27b73a65494" />

Nvim:
<img width="1112" alt="Screenshot 2025-05-26 at 14 35 41" src="https://github.com/user-attachments/assets/b0bc8360-698a-4f75-b94e-a2a39b9eb8db" />

Theme is Catppuccin Mocha

**Reply by mtalexan** [reply]
I think that's from a specific VSCode Extension, I don't see any such multi-color heading highlighting in mine.

**Reply by uncenter** [reply]
@mtalexan that would be because your theme doesn't make use of the TextMate scopes for each level. If you run "Developer: Inspect Editor Tokens and Scopes" in the VS Code command palette and click on different Markdown heading levels, you'll observe TextMate scopes such as "heading.1.markdown", "heading.2.markdown", etc.

Same thing for Helix and Neovim, both support specific highlight captures for each level of heading: 

https://github.com/helix-editor/helix/blob/8961ae1dc66633ea6c9f761896cb0d885ae078ed/runtime/queries/markdown/highlights.scm#L5-L10

https://github.com/nvim-treesitter/nvim-treesitter/blob/42fc28ba918343ebfd5565147a42a26580579482/queries/markdown/highlights.scm#L2-L26

**Reply by PaulRBerg** [reply]
came here to report the same issue - headers should be colored. It looks dull:

<img width="1326" height="410" alt="image" src="https://github.com/user-attachments/assets/8bf18386-b06e-4581-bdbc-7a1006817269" />

**Reply by Yu-Chieh-Henry-Yang** [reply]
Yes, coloring each level of header differently will make a big difference when reading the raw markdown file.

I have opened a dedicated discussion here: https://github.com/zed-industries/zed/discussions/58023

Please help upvote it to increase the chance of it being implemented if anyone reading this comment is interested in the feature too!

**Comment by mtalexan** [comment]
It's important to understand that in Markdown, there's a "standard" Markdown which is pretty limited and very poorly specified.  Often what's actually implemented for "standard" Markdown is [CommonMark Markdown](https://commonmark.org/), which is an actual strict definition for the previously only loosely defined "standard" Markdown format.  That's what most formatters and flavors are actually based on.  
Markdown flavors are actually custom syntax extensions to the base standard Markdown format, and there are a LOT of them.  One of the most popular is GitHub-flavored Markdown, but there are others and often different flavors may have conflicting syntax extensions.  
VSCode assumes GitHub-flavored markdown as their default, but currently Zed seems to pick "standard" Markdown as the default. Neither seems to have very good support for other Markdown flavors that I've seen.

On top of Markdown flavors, there are also some document writing features and knowledgebase ease-of-use features that aren't present in Markdown or any of it's major flavors.  Markdown has increasingly been getting adopted as an easy rendered document language for knowledge tools (e.g. Roam, Obsidian.md, Notion), and then as a quick Blog website source document format that can be generated from files like that, as well as a rendered research paper/whitepaper source format (as an alternative to Latex).  Resultantly there's increasing interest in supporting one-off syntax extensions to simplify use and implementation of some of these features rather than having to fall back to the raw HTML format (which Markdown also supports and allows normally). Some examples of these one-off features are Footnotes, captions, and custom reference anchors.  Being able to quickly create them, as well as being able to render them is a two-part issue.

**Comment by Yaume7** [comment]
I'm using Chatgpt for help coding, in general in OpenWebUI

Using vscode, a copy/paste from vscode to chatgpt automaticly put the code markdown format (the ``` before and after the code)

I'm using Zed now, and when i do the same, there is no markdown so it's not in a code format, it makes it more difficult to use, and copy paste back on Zed. On Openwebui, it make it bug the UI.
<img width="1009" height="304" alt="2025-09-19_00-09" src="https://github.com/user-attachments/assets/7b61d56b-e041-479e-8f20-1d838f99e57c" />
I've join a screen from the bug on open webui
If I add manually the ```, it has the good format.

**Comment by spamshaker** [comment]
Maybe it would be possible to implement sync scrolling between editor and renderer in split left/right mode?

**Comment by injust** [comment]
Text selection and copying (https://github.com/zed-industries/zed/discussions/47217) is a pretty big one for me.

**Comment by injust** [comment]
Going to add https://github.com/zed-industries/zed/discussions/47334 here as well, for copying links in Markdown Preview.

**Comment by dovelus** [comment]
I don't know if this was already reported but inlinde code block don't render with monospace font and are not at least following the standard on how they are displayed, like having a background with rounder corners or stuff like that. 
I was able to produce a fix locally and it took me like 20 minutes:
ORIGINAL:
<img width="1535" height="359" alt="image" src="https://github.com/user-attachments/assets/968bbb6e-bc0a-4ab6-90ae-eeafa76dfab9" />

MODIFIED:
<img width="1520" height="325" alt="image" src="https://github.com/user-attachments/assets/528236ba-3191-4850-bc9d-2046ef6a6f8b" />

**Reply by ian-h-chamberlain** [reply]
Yep, see https://github.com/zed-industries/zed/issues/44320

There was an attempt to fix it already: https://github.com/zed-industries/zed/pull/44764
but I guess the Zed team is trying to consolidate implementations between `markdown_preview` (which doesn't support this) and `markdown` (which does support it, I guess mostly used for stuff like agent panel and other Zed UI).

I wasn't able to find any other issues about the combining of the two crates (which would hopefully solve this!) but since this is a "tracking issue" discussion I wanted to just point it out.

**Comment by dive** [comment]
Perhaps it only bothers me because I do not see it in the list of improvements. However, it really bothers me that the markdown preview font size is directly linked to the editor font size. So, I would like to propose a small change here: #51511

**Comment by kanstantsin-bucha** [comment]
Can we modify the table to add a support of Columns width based on content?
Now it is hard to read tables in the preview
<img width="1152" height="248" alt="Screenshot 2026-05-14 at 11 42 33" src="https://github.com/user-attachments/assets/8fdc853c-5b8e-466f-8139-6d7513cb0295" />

**Reply by mosantis** [reply]
@kanstantsin-bucha Something like this?
<img width="1925" height="1393" alt="kanstantsin_dark" src="https://github.com/user-attachments/assets/d28ade17-a3d4-4a17-85e4-fab303821ec0" />

**Comment by mosantis** [comment]
Newbie here. I am trying to work on a Markdown "improvement" [on this thread](https://github.com/zed-industries/zed/discussions/43384) but not sure if I should post here. Thoughts?

**Comment by Sinthoras7** [comment]
It would be great, if there were an option to have in preview a soft-line-wrap for long lines in a fenced code block. (With the additional options to keep indentation and to insert a small symbol that shows, that there is a newline happening, as e.g. done with `set wrap`
`set breakindent` and `set showbreak=↳\` in neovim).
